designs/timeout-api.md

```diff
@@ -0,0 +1,78 @@
+---
+state: draft
+---
+
+# Node Timout API
+
+## Summary
+
+This design looks at how to add a timeout mechanism to nodes. This will allow
+flows to be created that can handle hangs or other unexpected events that prevent
+the normal operation of a flow.
+
+## Authors
+
+ - @knolleary
+
+## Details
+
+This design builds on the new [Messsaging API design](node-messaging-api) that
+introduces the `node.done` function for a node to indicate it has completed
+handling a message.
+
+It will be possible for the runtime to timeout a node that does not complete its
+processing of a message within a given time. But this can only work if the runtime
+knows a node has been updated to call `node.done`.
+
+**By default, a Node cannot be timed out**
+
+It will be up to individual implementations to decide if it makes sense to add
+timeout handling. Some nodes, such as the Change and Switch nodes do not perform
+any long-running activity, so would not make sense to enable timeout handling in
+them.
+
+To enable timeout handling a node must add an event handler for the `timeout` event:
+
+    this.on('timeout', function(msg) {
+        // Process this timeout. For example, cancel any in-flight work
+    })
+
+By adding this handler, the node is telling the runtime it can be timed out and,
+therefore, that the node *will* call `node.done(msg)` when it finishes with each
+message.
+
+The runtime will track the messages passed to the node using a `Set` - this
```

Comment:
I'm concerned about memory leak of `Set` caused by missing node.done call or error occurrence.
Can the messages be closure variable of message timeout handler?

Reply:
If a node has told us it is going to handle timeouts, then it is committing to calling `node.done` and being timed out.

If the node does call `done` then we can remove the message from the set.

If the node does not call `done` then it will timeout and we can remove the message from the set.

I did consider whether WeakSet would be useful here - as that will allow the message to be garbage collected. But it is not appropriate here because if we timeout the message and the GC has already collected it, we won't have the message available to pass to the timeout handling.